logstash-core/lib/logstash/instrument/periodic_poller/cgroup.rb

```diff
@@ -6,130 +6,234 @@
 module LogStash module Instrument module PeriodicPoller
   class Cgroup
     include LogStash::Util::Loggable
+    class Override
+      attr_reader :key, :value
+      def initialize(key)
+        @key = key
+        @value = java.lang.System.getProperty(@key)
+      end
 
-    CONTROL_GROUP_RE = Regexp.compile("\\d+:([^:,]+(?:,[^:,]+)?):(/.*)");
-    CONTROLLER_SEPARATOR_RE = ","
-
-    PROC_SELF_CGROUP_FILE = Pathname.new("/proc/self/cgroup")
-    PROC_CGROUP_CPU_DIR = Pathname.new("/sys/fs/cgroup/cpu")
-    PROC_CGROUP_CPUACCT_DIR = Pathname.new("/sys/fs/cgroup/cpuacct")
+      def nil?
+        value.nil?
+      end
 
-    GROUP_CPUACCT = "cpuacct"
-    CPUACCT_USAGE_FILE = "cpuacct.usage"
+      def override(other)
+        nil? ? other : value
+      end
+    end
 
-    GROUP_CPU = "cpu"
-    CPU_FS_PERIOD_US_FILE = "cpu.cfs_period_us"
-    CPU_FS_QUOTA_US_FILE = "cpu.cfs_quota_us"
+    class NullOverride
+      def override(other)
+        other
+      end
+    end
 
-    CPU_STATS_FILE = "cpu.stat"
+    ## `/proc/self/cgroup` contents look like this
+    # 5:cpu,cpuacct:/
+    # 4:cpuset:/
+    # 2:net_cls,net_prio:/
+    # 0::/user.slice/user-1000.slice/session-932.scope
+    ## e.g. N:controller:/path-to-info
+    # we find the controller and path
+    # we skip the line without a controller e.g. 0::/path
+    # we assume there are these symlinks:
+    # `/sys/fs/cgroup/cpu` -> `/sys/fs/cgroup/cpu,cpuacct
+    # `/sys/fs/cgroup/cpuacct` -> `/sys/fs/cgroup/cpu,cpuacct
+
+    class CGroupResources
+      attr_reader :cgroup_file, :cpu_dir, :cpuacct_dir
+      attr_reader :cpu_path_override, :cpuacct_path_override
+      attr_reader :controller_cpuacct, :controller_cpu
+
+      CONTROL_GROUP_RE = Regexp.compile("\\d+:([^:,]+(?:,[^:,]+)?):(/.*)")
+      CONTROLLER_SEPARATOR_RE = ","
+
+      def initialize
```

Comment:
All of the statically-defined `Pathname`s could be constants here, since `Pathname` is immutable.